File: day01.py
#!/usr/bin/python3

################################### part 1 ###################################

group1 = []
group2 = []

with open("input.txt") as file: 
    for line in file:
        group1.append(int(line.split()[0]))
        group2.append(int(line.split()[1]))

group1.sort()
group2.sort()

total_dist = 0

for i in range(0,len(group1)):
    dist = abs( group1[i] - group2[i] )
    total_dist += dist
    #print(i, group1[i], group2[i], dist, total_dist)

print("Total distance between the two lists:", total_dist)

################################### part 2 ###################################

similarity = [loc_id * group2.count(loc_id) for loc_id in group1]
total_sim = sum(similarity)

print("Similarity score of the two lists:", total_sim)

File: day01.js
const fs = require("node:fs");

function calcDistance (arrayOne, arrayTwo) {

    const group1Sorted = arrayOne.sort();
    const group2Sorted = arrayTwo.sort();
    
    const distances = group1Sorted.map((locID, i) => Math.abs( locID - group2Sorted[i]));
  
    return distances.reduce((totalDist, dist) => totalDist + dist);
}

function calcSimilarity (arrayOne, arrayTwo) {

    const similarity = arrayOne.map((locID1) => 
        locID1 * arrayTwo.reduce((countID, locID2) => 
            (locID2 === locID1 ? countID + 1 : countID), 0)
    );

    return similarity.reduce((totalSim, sim) => totalSim + sim);
}

fs.readFile("input.txt", "utf8", (err, data) => {
    if (err) {
        console.error(err);
        return;
    } 
    
    const lines = data.split('\n');

    const group1 = [];
    const group2 = [];

    for (line of lines) {
        group1.push(Number(line.split(/\s/).filter(n => n)[0]));
        group2.push(Number(line.split(/\s/).filter(n => n)[1]));
    }

    // part 1
    console.log("Total distance between the two lists:", calcDistance(group1, group2));

    // part 2
    console.log("Similarity score of the two lists:", calcSimilarity(group1, group2));
});
